lib/chai/core/assertions.js

```diff
@@ -1268,120 +1268,114 @@ module.exports = function (chai, _) {
-      // first, check desired error
-      if (desiredError) {
-        this.assert(
-            err === desiredError
-          , 'expected #{this} to throw #{exp} but #{act} was thrown'
-          , 'expected #{this} to not throw #{exp}'
-          , (desiredError instanceof Error ? desiredError.toString() : desiredError)
-          , (err instanceof Error ? err.toString() : err)
-        );
-
-        flag(this, 'object', err);
-        return this;
+    function similarException (expected, thrown) {
+      var exception = expected.e;
+      if (exception && exception.constructor === thrown.constructor) {
+        return exception.name === thrown.name;
       }
+      return !expected.type || thrown instanceof expected.type;
+    }
 
-      // next, check constructor
-      if (constructor) {
-        this.assert(
-            err instanceof constructor
-          , 'expected #{this} to throw #{exp} but #{act} was thrown'
-          , 'expected #{this} to not throw #{exp} but #{act} was thrown'
-          , name
-          , (err instanceof Error ? err.toString() : err)
-        );
-
-        if (!errMsg) {
-          flag(this, 'object', err);
-          return this;
-        }
+    function similarMessage (expected, thrown) {
+      var isThrownedException = typeof thrown === 'object' && thrown.hasOwnProperty('message');
+      var message = isThrownedException ? thrown.message : '' + thrown;
+      if (expected.errMsg instanceof RegExp) {
+        return expected.errMsg.test(message);
       }
+      if (!expected.errMsg) {
+        return true;
+      }
+      if (expected.action === 'matching') {
+        return message === expected.errMsg;
+      }
+      return !!~message.indexOf(expected.errMsg);
+    }
 
-      // next, check message
-      var message = 'error' === _.type(err) && "message" in err
-        ? err.message
-        : '' + err;
-
-      if ((message != null) && errMsg && errMsg instanceof RegExp) {
-        this.assert(
-            errMsg.exec(message)
-          , 'expected #{this} to throw error matching #{exp} but got #{act}'
-          , 'expected #{this} to throw error not matching #{exp}'
-          , errMsg
-          , message
-        );
-
-        flag(this, 'object', err);
-        return this;
-      } else if ((message != null) && errMsg && 'string' === typeof errMsg) {
-        this.assert(
-            ~message.indexOf(errMsg)
-          , 'expected #{this} to throw error including #{exp} but got #{act}'
-          , 'expected #{this} to throw error not including #{act}'
-          , errMsg
-          , message
-        );
-
-        flag(this, 'object', err);
-        return this;
+    function parseParams (constructor, errMsg) {
+      var expected = { placeholder: 'an error', action: 'matching' };
+      if (!arguments.length || !constructor) {
+        return expected;
+      }
+      if (constructor instanceof RegExp) {
+        expected.errMsg = constructor;
+        return expected;
+      }
```

Comment:
Well, in my humble opinion it would be better if we used `utils` to determine type, this avoids multiple parts of the code doing the same thing and therefore decreases the possibility of an error, because `utils` centralizes this kind of functionality so it is the only place an error can happen (instead of anywhere in the code).